finalfusion-ffi/src/lib.rs

```diff
@@ -75,6 +75,39 @@ pub unsafe extern "C" fn ff_free_embeddings(embeddings: *mut Embeddings<VocabWra
     }
 }
 
+/// Write the embeddings.
+///
+/// Writes the embeddings to the given `filename`.
+///
+/// Returns:
+///  * -1 if the file could not be created or writing failed.
+///  * 0 if the file was written succesfully.
+#[no_mangle]
+pub unsafe extern "C" fn ff_write_embeddings(
+    embeddings: *mut Embeddings<VocabWrap, StorageWrap>,
+    filename: *const c_char,
+) -> isize {
+    check_null!(embeddings);
+    check_null!(filename);
+    let embeddings = &*embeddings;
+    let filename = CStr::from_ptr(filename);
+    let filename = OsStr::from_bytes(filename.to_bytes());
+    let mut writer = match File::create(filename) {
+        Ok(f) => BufWriter::new(f),
+        Err(err) => {
+            update_error(&format!("{}", err));
```

Comment:
These io errors are usually very terse (don't have a filename). Maybe change the formatting string to something like, `"Could not open file {} for writing: {}"`